temporal-sdk/src/main/java/io/temporal/failure/DefaultFailureConverter.java

```diff
@@ -0,0 +1,383 @@
+import io.temporal.api.failure.v1.CanceledFailureInfo;
+import io.temporal.api.failure.v1.ChildWorkflowExecutionFailureInfo;
+import io.temporal.api.failure.v1.Failure;
+import io.temporal.api.failure.v1.ResetWorkflowFailureInfo;
+import io.temporal.api.failure.v1.ServerFailureInfo;
+import io.temporal.api.failure.v1.TerminatedFailureInfo;
+import io.temporal.api.failure.v1.TimeoutFailureInfo;
+import io.temporal.client.ActivityCanceledException;
+import io.temporal.common.converter.DataConverter;
+import io.temporal.common.converter.DefaultDataConverter;
+import io.temporal.common.converter.EncodedValues;
+import io.temporal.serviceclient.CheckedExceptionWrapper;
+import java.io.PrintWriter;
+import java.io.StringWriter;
+import java.lang.reflect.Type;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class DefaultFailureConverter implements FailureConverter {
+
+  private static final Logger log = LoggerFactory.getLogger(DefaultFailureConverter.class);
+
+  public static final String JAVA_SDK = "JavaSDK";
+
+  private static final String ENCODED_FAILURE_MESSAGE = "Encoded failure";
+  private static final String STACK_TRACE_KEY = "stack_trace";
+  private static final String MESSAGE_KEY = "message";
+
+  private static final Type HASH_MAP_STRING_STRING_TYPE =
+      new TypeToken<HashMap<String, String>>() {}.getType();
+
+  /**
+   * Stop emitting stack trace after this line. Makes serialized stack traces more readable and
+   * compact as it omits most of framework level code.
+   */
+  private static final ImmutableSet<String> CUTOFF_METHOD_NAMES =
+      ImmutableSet.of(
+          "io.temporal.internal.worker.POJOActivityImplementationFactory$POJOActivityImplementation.execute",
+          "io.temporal.internal.sync.POJOWorkflowTaskHandler$POJOWorkflowImplementation.execute");
+
+  /** Used to parse a stack trace line. */
+  private static final String TRACE_ELEMENT_REGEXP =
+      "((?<className>.*)\\.(?<methodName>.*))\\(((?<fileName>.*?)(:(?<lineNumber>\\d+))?)\\)";
+
+  private static final Pattern TRACE_ELEMENT_PATTERN = Pattern.compile(TRACE_ELEMENT_REGEXP);
+
+  private boolean encodeDefaultAttributes;
+
+  public DefaultFailureConverter() {
+    this(false);
+  }
+
+  public DefaultFailureConverter(boolean encodeDefaultAttributes) {
+    this.encodeDefaultAttributes = encodeDefaultAttributes;
+  }
+
+  @Override
+  public RuntimeException failureToException(Failure failure, DataConverter dataConverter) {
+    if (failure == null) {
+      return null;
+    }
+    if (failure.hasEncodedAttributes()) {
+      Payload encodedAttributesPayload = failure.getEncodedAttributes();
+      Map<String, String> encodedAttributes =
+          DefaultDataConverter.STANDARD_INSTANCE.fromPayload(
+              encodedAttributesPayload, HashMap.class, HASH_MAP_STRING_STRING_TYPE);
```

Comment:
I think you should use a static inner pojo that represents this attribute as a Jackson type for clearer conversion.